File: finance/server/shared/consent_ledger.py
# -*- coding: utf-8 -*-
"""동의 원장: 부여·철회·만료를 append-only로 기록 (AC-5)."""
from dataclasses import dataclass


@dataclass(frozen=True)
class ConsentRecord:
    user_id: str
    scope: str
    status: str  # granted | withdrawn | expired
    seq: int


class ConsentLedger:
    """append-only 동의 원장. 최신 레코드가 현재 상태를 결정한다."""

    def __init__(self):
        self._records = []

    def _append(self, user_id, scope, status):
        rec = ConsentRecord(user_id, scope, status, len(self._records) + 1)
        self._records.append(rec)
        return rec

    def grant(self, user_id, scope="mydata"):
        return self._append(user_id, scope, "granted")

    def withdraw(self, user_id, scope="mydata"):
        return self._append(user_id, scope, "withdrawn")

    def expire(self, user_id, scope="mydata"):
        return self._append(user_id, scope, "expired")

    def is_active(self, user_id, scope="mydata"):
        latest = None
        for r in self._records:
            if r.user_id == user_id and r.scope == scope:
                latest = r
        return latest is not None and latest.status == "granted"

    def records(self):
        return list(self._records)

    def count(self):
        return len(self._records)

File: finance/java/complete/src/main/java/com/datasense/finance/repository/ConsentLedger.java
package com.datasense.finance.repository;

import com.datasense.finance.domain.ConsentRecord;
import org.springframework.stereotype.Repository;

import java.util.ArrayList;
import java.util.List;

/**
 * 동의 원장: 부여·철회·만료를 append-only로 기록 (AC-5).
 * 최신 레코드가 현재 상태를 결정한다. (shared/consent_ledger.py 포팅, 인메모리)
 */
@Repository
public class ConsentLedger {

    private final List<ConsentRecord> records = new ArrayList<>();

    private synchronized ConsentRecord append(String userId, String scope, String status) {
        ConsentRecord rec = new ConsentRecord(userId, scope, status, records.size() + 1);
        records.add(rec);
        return rec;
    }

    public ConsentRecord grant(String userId, String scope) {
        return append(userId, scope, "granted");
    }

    public ConsentRecord withdraw(String userId, String scope) {
        return append(userId, scope, "withdrawn");
    }

    public ConsentRecord expire(String userId, String scope) {
        return append(userId, scope, "expired");
    }

    /** 최신 레코드가 granted면 활성. */
    public synchronized boolean isActive(String userId, String scope) {
        ConsentRecord latest = null;
        for (ConsentRecord r : records) {
            if (r.userId().equals(userId) && r.scope().equals(scope)) {
                latest = r;
            }
        }
        return latest != null && "granted".equals(latest.status());
    }

    public synchronized List<ConsentRecord> records() {
        return new ArrayList<>(records);
    }

    public synchronized int count() {
        return records.size();
    }
}
